java/ql/src/experimental/Security/CWE/CWE-706/UserControlledArbitraryRead.ql

```diff
@@ -0,0 +1,221 @@
+class TypeServiceResponse extends Class {
+  TypeServiceResponse() { this.hasQualifiedName("ai.susi.server", "ServiceResponse") }
+}
+
+class ServiceResponseSink extends DataFlow::ExprNode {
+  ServiceResponseSink() {
+    exists(ConstructorCall call |
+      call.getConstructedType() instanceof TypeServiceResponse and
+      this.getExpr() = call.getAnArgument()
+    )
+    or
+    exists(MethodAccess call |
+      call.getType() instanceof TypeServiceResponse and
+      this.getExpr() = call.getAnArgument()
+    )
+  }
+}
+
+predicate deletesFile(DataFlow::ExprNode node) {
+  exists(MethodAccess call |
+    call.getReceiverType() instanceof TypeFile and
+    call.getMethod().getName().matches("delete%") and
+    node.getExpr() = call.getQualifier()
+  )
+}
+
+predicate deletesPath(DataFlow::ExprNode node) {
+  exists(MethodAccess call |
+    call.getReceiverType() instanceof TypeFiles and
+    call.getMethod().getName().matches("delete%") and
+    node.getExpr() = call.getArgument(0)
+  )
+}
+
+predicate renamesFile(DataFlow::ExprNode node) {
+  exists(MethodAccess call |
+    call.getReceiverType() instanceof TypeFile and
+    call.getMethod().getName().matches("renameTo%") and
+    (
+      node.getExpr() = call.getQualifier()
+      or
+      node.getExpr() = call.getArgument(0)
+    )
+  )
+}
+
+predicate renamesPath(DataFlow::ExprNode node) {
+  exists(MethodAccess call |
+    call.getReceiverType() instanceof TypeFiles and
+    call.getMethod().getName().matches("move%") and
+    (
+      node.getExpr() = call.getArgument(0)
+      or
+      node.getExpr() = call.getArgument(1)
+    )
+  )
+}
+
+class SensitiveFileOperationSink extends DataFlow::ExprNode {
+  SensitiveFileOperationSink() {
+    deletesFile(this)
+    or
+    deletesPath(this)
+    or
+    renamesFile(this)
+    or
+    renamesPath(this)
+  }
+}
+
+/**    Holds if `node1` is used in the creation of `node2` and not guarded. */
+predicate usedInPathCreation(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(Expr e | e = node1.asExpr() |
+    e = node2.asExpr().(PathCreation).getInput() and not guarded(e)
+  )
+}
+
+predicate putsValueIntoJsonObject(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess call |
+    call.getReceiverType() instanceof TypeJsonObject and
+    call.getMethod().getName() = ["put", "putOnce", "putOpt"] and
+    call.getQualifier() = node2.asExpr() and
+    call.getArgument(1) = node1.asExpr()
+  )
+}
+
+predicate putsValueIntoJsonArray(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess call |
+    call.getReceiverType() instanceof TypeJsonArray and
+    call.getMethod().getName() = "put" and
+    call.getQualifier() = node2.asExpr() and
+    (
+      call.getArgument(1) = node1.asExpr() and call.getNumArgument() = 2
+      or
+      call.getArgument(0) = node1.asExpr() and call.getNumArgument() = 1
+    )
+  )
+}
+
+class ContainsDotDotSanitizer extends DataFlow::BarrierGuard {
```

Comment:
This might cause false negatives, e.g.
```java
String s = ...;
Path ROOT = Path.of("/my-app/data);
if (!s.contains("..")) {
  Files.delete(ROOT.resolve(Path.of(s));
  // Or
  Files.delete(ROOT.resolve(s));
)
```
Would still be vulnerable if `s` is absolute.